pkg/controllermanager/controller/secretbinding/secretbinding_control.go

```diff
@@ -135,19 +137,27 @@ func (c *defaultControl) ReconcileSecretBinding(obj *gardencorev1beta1.SecretBin
 		if len(associatedShoots) == 0 {
 			secretBindingLogger.Info("No Shoots are referencing the SecretBinding. Deletion accepted.")
 
-			// Remove finalizer from referenced secret
-			secret, err := c.secretLister.Secrets(secretBinding.SecretRef.Namespace).Get(secretBinding.SecretRef.Name)
-			if err == nil {
-				secretFinalizers := sets.NewString(secret.Finalizers...)
-				secretFinalizers.Delete(gardencorev1beta1.ExternalGardenerName)
-				secret.Finalizers = secretFinalizers.UnsortedList()
-				if err := c.k8sGardenClient.Client().Update(ctx, secret); client.IgnoreNotFound(err) != nil {
+			mayReleaseSecret, err := c.mayReleaseSecret(secretBinding.SecretRef.Namespace, secretBinding.SecretRef.Name)
+			if err != nil {
+				secretBindingLogger.Error(err.Error())
+				return err
+			}
+
+			if mayReleaseSecret {
+				// Remove finalizer from referenced secret
+				secret, err := c.secretLister.Secrets(secretBinding.SecretRef.Namespace).Get(secretBinding.SecretRef.Name)
+				if err == nil {
+					secretFinalizers := sets.NewString(secret.Finalizers...)
```

Comment:
Shouldn't we use .DeepCopy for things we get from listers?
Btw there is a RemoveFinalizer/RemoveFinalizer func. Does it makes sense to use it?

Reply:
I didn't change this logic as part of this PR, but I can quickly refactor it.

Reply:
Ah, yep, I guess we can do it with another PR as we would need to touch some other parts.

Reply:
np, I'm doing it quickly, one sec.